File: 2021/Day 13/C++/part2.cpp
#include <fstream>
#include <iostream>
#include <map>
#include <sstream>
#include <vector>

// Namespaces
using namespace std;  // Bad practice, I know

inline istream& operator>>(istream& is, const char* checkstr);

// Main
int main() {
    // Get input
    ifstream file("../data.txt");
    map<pair<int, int>, bool> paper = {};
    vector<pair<char, int>> folds;
    string line;
    while (getline(file, line)) {
        if (line.empty()) continue;
        istringstream is(line);
        if (line[0] == 'f') {
            pair<char, int> p;
            is >> "fold along " >> p.first >> "=" >> p.second;
            folds.push_back(p);
        } else {
            pair<int, int> p;
            is >> p.first >> "," >> p.second;
            paper[p] = true;
        }
    }

    // Fold paper
    for (size_t i = 0; i < folds.size(); i++) {
        char axis = folds[i].first;
        int index = folds[i].second;
        for (auto it = paper.cbegin(); it != paper.cend();) {
            auto dot = it->first;
            if (axis == 'y' && dot.second > index) {
                if (dot.second > index) {
                    pair<int, int> p = {dot.first, 2 * index - int(dot.second)};
                    paper[p] = true;         // Mirror dot
                    it = paper.erase(it++);  // Delete dot
                    continue;
                }
            } else if (axis == 'x' && dot.first > index) {
                pair<int, int> p = {2 * index - dot.first, dot.second};
                paper[p] = true;         // Mirror dot
                it = paper.erase(it++);  // Delete dot
                continue;
            }
            ++it;
        }
    }

    // Print dots
    int xmax = 0, ymax = 0;
    for (auto dotPair : paper) {
        if (dotPair.first.first > xmax) xmax = dotPair.first.first;
        if (dotPair.first.second > ymax) ymax = dotPair.first.second;
    }
    for (int y = 0; y <= ymax; y++) {
        for (int x = 0; x <= xmax; x++) {
            if (paper.count(pair<int, int>{x, y}))
                cout << '#';
            else
                cout << ' ';
        }
        cout << endl;
    }
    return 0;
}

// read a delimiter that must match a certain string
// eg: is >> x >> "," >> y;
inline istream& operator>>(istream& is, const char* checkstr) {
    // skip ws at the start of checkstr
    const char* currcheck = checkstr;
    while (isspace(*currcheck)) ++currcheck;

    // skip ws at start of istream
    while (isspace(is.peek())) is.get();

    // go through checkstr char by char and ensure we get matching data out of is
    while (*currcheck) {
        is.get();
        ++currcheck;
    }

    return is;
}

File: 2021/Day 13/TS/part2.ts
import fs from 'fs'
import path from 'path'

const file: string = fs.readFileSync(path.resolve(__dirname, '../data.txt'), 'utf-8')
const parts = file.split('\n\n').map((part) => part.split('\n'))
const dots = parts[0].map((line) => line.split(',').map((num) => parseInt(num)))
const directions = parts[1].map((line) => line.split('fold along ')[1].split('='))

// Map the dots
const paper: { [key: string]: { [key: string]: boolean } } = {}
for (let [x, y] of dots) {
    if (!paper[y]) paper[y] = {}
    paper[y][x] = true
}

// Fold paper
for (let direction of directions) {
    const axis = direction[0]
    const index = Number(direction[1])
    if (axis == 'y') {
        for (let key of Object.keys(paper)) {
            const y = Number(key)
            if (y > index) {
                for (let x of Object.keys(paper[y])) {
                    if (!paper[2 * index - y]) paper[2 * index - y] = {}
                    paper[2 * index - y][x] = true // Mirror folded dots
                }
                delete paper[y] // Delete folded dots
            }
        }
        delete paper[index] // Delete fold line
    } else {
        for (let [ykey, xmap] of Object.entries(paper)) {
            const y = Number(ykey)
            for (let xkey of Object.keys(xmap)) {
                const x = Number(xkey)
                if (x >= index) {
                    xmap[2 * index - x] = true // Mirror folded dots
                    delete xmap[x] // Delete folded dot
                }
            }
        }
    }
}

// Print dots
console.log('Answer:\n')
const ymax = Number(Object.keys(paper).reduce((max, val) => (val > max ? val : max)))
const xmax = Object.values(paper)
    .map((line) => Object.keys(line).reduce((max, val) => (Number(val) > max ? Number(val) : max), 0))
    .reduce((max, val) => (val > max ? val : max))
for (let y = 0; y <= ymax; y++) {
    for (let x = 0; x <= xmax; x++) {
        process.stdout.write(paper?.[y]?.[x] ? '#' : ' ')
    }
    console.log()
}
